File: src/ackward/logging/Formatter.cpp
#include <ackward/logging/Formatter.hpp>

#include <boost/python.hpp>

#include <ackward/logging/LogRecord.hpp>
#include <ackward/logging/Module.hpp>

using namespace boost::python;

namespace ackward { namespace logging
{

Formatter::Formatter() :
    Object(
        module().attr("Formatter")())
{}

Formatter::Formatter(boost::python::object o) :
    Object (o)
{}

std::wstring Formatter::format(const LogRecord& r) const
{
    return extract<std::wstring>(
        obj().attr("format")(r));
}

std::wstring Formatter::formatTime(const LogRecord& r) const
{
    return extract<std::wstring>(
        obj().attr("formatTime")(r));
}

std::wstring Formatter::formatTime(const LogRecord& r, 
                                   const std::wstring& datefmt) const
{
    return extract<std::wstring>(
        obj().attr("formatTime")(r, datefmt));
}

std::wstring Formatter::formatException(boost::python::tuple excInfo) const
{
    return extract<std::wstring>(
        obj().attr("formatException")(excInfo));
}

}}

File: src/ackward/logging/Formatter.py
import functools, operator

from ackward import (Class,
                     Constructor,
                     method,
                     Namespace,
                     TranslationUnit)

formatter_doc = '''\\rst
Wrapper around ``logging.Formatter`` objects. See `<http://docs.python.org/py3k/library/logging.html#formatter-objects>`_ for more information.
\\endrst'''

def tunit():
    return TranslationUnit(
        guard='INCLUDE_ACKWARD_LOGGING_FORMATTER',
        forward_declarations=[
            ('ackward', 'logging', 'class LogRecord'),
            ('boost', 'python', 'class tuple'),
            ],
        header_includes=[
            ('string',),
            ],
        impl_includes=[
            ('ackward', 'logging', 'Formatter.hpp'),
            ('ackward', 'logging', 'LogRecord.hpp'),
            ('boost', 'python', 'tuple.hpp'),
            ])

def methods(parent):
    m = [
        'std::wstring format(LogRecord record) const',
        'std::wstring formatTime(LogRecord record) const',
        'std::wstring formatTime(LogRecord record, std::wstring datefmt) const',
        'std::wstring formatException(boost::python::tuple exc) const'
        ]
    list(map(lambda m: method(m, parent=parent), m))

def definition(env):
    t = tunit()
    ns = Namespace('ackward', 'logging', parent=t)
    cls = Class(name='Formatter',
                wrapped_class='logging.Formatter',
                parent=ns,
                doc=formatter_doc)
    Constructor(parent=cls)
    methods(parent=cls)

    return t
